Nordigen Custom agreements (wrong branch)

dshoreman opened the pull request:
Resubmit of #10425 in light of #10501.

Needs rebase again before I can test/fix anything but I'll take care of that tomorrow.

dshoreman committed:
Nordigen: Pass custom agreement with max history

In order to change the total days of transactions we can request, a
custom end user agreement is needed. There *are* methods to find
existing agreements, but assuming nothing else breaks it's probably
fairly safe to just go ahead and create one.

If it works, it only runs once... as far as I can tell.

dshoreman committed:
Nordigen: Pass custom agreement with max history

In order to change the total days of transactions we can request, a
custom end user agreement is needed. There *are* methods to find
existing agreements, but assuming nothing else breaks it's probably
fairly safe to just go ahead and create one.

If it works, it only runs once... as far as I can tell.

dshoreman committed:
Nordigen: Use institution max for bank integration

While we set the days in the agreement, how many we fetch when
processing transactions depends on the `from_date` of the bank
integration. We could call `getInstitution($id)` on the API, but
consistency would dictate adding a wrapper in the Nordigen class and
it's already being called within `$nordigen->getAccount()` so it makes
more sense to extract the data in the account transformer instead.

App\Jobs\Bank\ProcessBankTransactionsNordigen also sets a from date, but
that's only used when not set on the bank integration so it can be left
at the Nordigen default of 90 days worth of transaction history.

dshoreman committed:
Nordigen: Use institution max for bank integration

While we set the days in the agreement, how many we fetch when
processing transactions depends on the `from_date` of the bank
integration. We could call `getInstitution($id)` on the API, but
consistency would dictate adding a wrapper in the Nordigen class and
it's already being called within `$nordigen->getAccount()` so it makes
more sense to extract the data in the account transformer instead.

App\Jobs\Bank\ProcessBankTransactionsNordigen also sets a from date, but
that's only used when not set on the bank integration so it can be left
at the Nordigen default of 90 days worth of transaction history.

dshoreman committed:
Enable country filtering for Nordigen bank list

Enabling the country filter stops the selector loop from working because
instead of a `.ob-list-institution` it starts with `.ob-country-list`
items which then get replaced on click so banks don't exist on load.

By wrapping the institution loop in an observer it'll automatically
attach the events when a country is selected and it works wonderfully.

dshoreman committed:
Enable country filtering for Nordigen bank list

Enabling the country filter stops the selector loop from working because
instead of a `.ob-list-institution` it starts with `.ob-country-list`
items which then get replaced on click so banks don't exist on load.

By wrapping the institution loop in an observer it'll automatically
attach the events when a country is selected and it works wonderfully.

dshoreman committed:
Allow setting number of days of transactions

If e.g. `&tx_days=365` is added to the URL, this will use it to set the
amount of history that's fetched from Nordigen on sync, with basic
safeguards to prevent it being set too low or above the bank's max.

dshoreman committed:
Allow setting number of days of transactions

If e.g. `&tx_days=365` is added to the URL, this will use it to set the
amount of history that's fetched from Nordigen on sync, with basic
safeguards to prevent it being set too low or above the bank's max.

dshoreman committed:
Add selection screen for transaction history

Instead of directing the user straight to GoCardless, when a bank is
selected this will replace the institution list with a list of options
from 30 days—up to the bank's transaction limit—in 30 day increments.

dshoreman committed:
Add selection screen for transaction history

Instead of directing the user straight to GoCardless, when a bank is
selected this will replace the institution list with a list of options
from 30 days—up to the bank's transaction limit—in 30 day increments.

dshoreman committed:
Improve Nordigen error-handling code

All failures boil down to about 3 variations - one with reason/lang, one
that has context defined but no company data, and a full variant that
also fills the company and account keys in the view.

The only other difference is redirect url coming from config when it's
not available in the context, so to simplify things this commit merges
all the failure view responses into one unified `failed` method.

dshoreman committed:
Improve Nordigen error-handling code

All failures boil down to about 3 variations - one with reason/lang, one
that has context defined but no company data, and a full variant that
also fills the company and account keys in the view.

The only other difference is redirect url coming from config when it's
not available in the context, so to simplify things this commit merges
all the failure view responses into one unified `failed` method.

dshoreman committed:
Fine-tune transaction history options

Replaces 360 days with 365, skips the last auto-generated option to
avoid getting e.g. 720 and 730 as options when it's close to max, and
gradually increases the accumulator so options aren't so overwhelming.

Since the loop will no longer generate the final option (0 < 15), this
commit also removes the `% 30` check so the real max is always used.

dshoreman committed:
Fine-tune transaction history options

Replaces 360 days with 365, skips the last auto-generated option to
avoid getting e.g. 720 and 730 as options when it's close to max, and
gradually increases the accumulator so options aren't so overwhelming.

Since the loop will no longer generate the final option (0 < 15), this
commit also removes the `% 30` check so the real max is always used.

dshoreman committed:
Fix back/close button display in Nordigen modal

When the country filtering mode is enabled, it adds a "Go back" button
but the close button is right on top of it. This moves the close button
to the right to avoid any click conflicts.

To fix an annoying display bug with the top offset, an observer is used
to override the default `display: none` of the back button to keep its
flex mode but with a hidden visibility so the UI doesn't move around.

dshoreman committed:
Fix back/close button display in Nordigen modal

When the country filtering mode is enabled, it adds a "Go back" button
but the close button is right on top of it. This moves the close button
to the right to avoid any click conflicts.

To fix an annoying display bug with the top offset, an observer is used
to override the default `display: none` of the back button to keep its
flex mode but with a hidden visibility so the UI doesn't move around.

dshoreman committed:
Find old EUA to get tx_days for new agreement

The end user agreement ID isn't stored with the bank integration, but it
*is* returned as part of the error for an expired account which works
perfectly for the case of renewing an expired requisition.

When `tx_days` isn't set in the request (i.e. it's a renewal) we instead
extract the EUA ID from the account error after getting the integration,
then once we have the EUA we're able to restore the old tx_days setting.

Since the BankIntegration query is used in both endpoints, this moves it
to a method with `firstOrFail()` rather than `first()` which also allows
for a cleanup of the integration saving code with try/catch/finally to
make it a little clearer which values apply to both new+existing cases.

dshoreman committed:
Find old EUA to get tx_days for new agreement

The end user agreement ID isn't stored with the bank integration, but it
*is* returned as part of the error for an expired account which works
perfectly for the case of renewing an expired requisition.

When `tx_days` isn't set in the request (i.e. it's a renewal) we instead
extract the EUA ID from the account error after getting the integration,
then once we have the EUA we're able to restore the old tx_days setting.

Since the BankIntegration query is used in both endpoints, this moves it
to a method with `firstOrFail()` rather than `first()` which also allows
for a cleanup of the integration saving code with try/catch/finally to
make it a little clearer which values apply to both new+existing cases.

dshoreman committed:
Fix error on token-invalid failure when expired

The cache data is set to expire after an hour, at which point the call
to `$request->getCompany()` will fail because it depends internally on
the `company_key` in the response of `getTokenContent()` which is null.

By moving the company *after* the context check, we avoid the exception
but need to manually set lang as there's no context for company data.

dshoreman committed:
Fix error on token-invalid failure when expired

The cache data is set to expire after an hour, at which point the call
to `$request->getCompany()` will fail because it depends internally on
the `company_key` in the response of `getTokenContent()` which is null.

By moving the company *after* the context check, we avoid the exception
but need to manually set lang as there's no context for company data.

dshoreman committed:
Cleanup, add return types, normalise quotes

dshoreman committed:
Cleanup, add return types, normalise quotes

dshoreman committed:
Attempt to reuse existing agreements

Implements changes from #10410 but using `Arr::first()` rather than a
foreach loop to fix bugs returning invalid or expired agreements.

If an agreement exists with at least the requested `$txDays` then that
is used, otherwise a new one is created with the given parameters. If it
fails, we error out because `createRequisition()` would fail regardless.

Skips accepting EUAs: seems it's done automatically during requisition.

dshoreman committed:
Attempt to reuse existing agreements

Implements changes from #10410 but using `Arr::first()` rather than a
foreach loop to fix bugs returning invalid or expired agreements.

If an agreement exists with at least the requested `$txDays` then that
is used, otherwise a new one is created with the given parameters. If it
fails, we error out because `createRequisition()` would fail regardless.

Skips accepting EUAs: seems it's done automatically during requisition.

dshoreman committed:
Support setting custom Nordigen access duration

If agreements are already available and no custom `access_days` is
passed, it will simply take the first match regardless of duration.

When a new agreement is needed, it defaults to the max supported.

dshoreman committed:
Support setting custom Nordigen access duration

If agreements are already available and no custom `access_days` is
passed, it will simply take the first match regardless of duration.

When a new agreement is needed, it defaults to the max supported.

dshoreman committed:
Show history selection screen instead of guessing

Removes all the guesswork for requisition renewals, instead opting for
the manual approach of simply redirecting to the form and updating its
handling such that it jumps straight to history selection when needed.

When the institution is preloaded (so we skip country/bank selection)
the cloned element is a country and not the bank so we need to replace
athe flag with its logo, but otherwise the code was all moved as-is.

dshoreman committed:
Show history selection screen instead of guessing

Removes all the guesswork for requisition renewals, instead opting for
the manual approach of simply redirecting to the form and updating its
handling such that it jumps straight to history selection when needed.

When the institution is preloaded (so we skip country/bank selection)
the cloned element is a country and not the bank so we need to replace
athe flag with its logo, but otherwise the code was all moved as-is.

dshoreman committed:
Add EUA error texts

dshoreman committed:
Add EUA error texts

turbo124 committed:
Refactor for html purifier

turbo124 committed:
Fixes for purifying html

turbo124 committed:
Fixes for statements

turbo124 committed:
Fixes for purifying html

turbo124 committed:
Clean up for static analysis

turbo124 committed:
Clean up for static analysis

turbo124 committed:
Clean up for static analysis

turbo124 committed:
Minor fixes for purify

turbo124 committed:
Logging in gh actions

turbo124 committed:
Logging in gh actions

turbo124 committed:
Logging in gh actions

turbo124 committed:
Change PDF engine for tests

turbo124 committed:
Fixes for calculate taxes

turbo124 committed:
v5.11.9

turbo124 committed:
Merge pull request #10463 from turbo124/v5-develop

v5.11.9

turbo124 committed:
Fixes for wrong keys

dshoreman committed:
Remove redundant `$expiresAt`

Leftover from a version that could use accepted EUAs.

If this functionality is ever restored, the `$eua['accepted'] === null`
should become `!$eua['accepted'] || $expiresAt > new DateTime('now')`.`

dshoreman committed:
Remove redundant `$expiresAt`

Leftover from a version that could use accepted EUAs.

If this functionality is ever restored, the `$eua['accepted'] === null`
should become `!$eua['accepted'] || $expiresAt > new DateTime('now')`.`

turbo124 committed:
Fixes for snappdf in container

turbo124 committed:
v5.11.10

turbo124 committed:
Merge pull request #10470 from turbo124/v5-develop

v5.11.10

turbo124 committed:
working on new design flow

benbrummer committed:
Remove "Prepare Laravel Application"

benbrummer committed:
Remove "Copy .env file"

benbrummer committed:
bootstrap/cache/* does not exist

benbrummer committed:
ensure to have correct permissions in the archive

benbrummer committed:
should work without sudo

benbrummer committed:
exclude for "./htaccess" has no match  and .htaccess should be part of the archive

benbrummer committed:
extension should match the archive type

turbo124 committed:
Minor fixes

turbo124 committed:
Additional handling for compressed PDFs

turbo124 committed:
Updates for designs and swiss qr code layout

turbo124 committed:
Refactor calm design

turbo124 committed:
Refactor calm design

turbo124 committed:
Refactor calm design

turbo124 committed:
Design migration

turbo124 committed:
v5.11.11

turbo124 committed:
Merge pull request #10480 from turbo124/v5-develop

v5.11.11

turbo124 committed:
Fixes for tests

turbo124 committed:
Minor fixes

turbo124 committed:
Merge pull request #10481 from turbo124/v5-develop

Minor fixes

turbo124 committed:
Fixes for property overloading

turbo124 committed:
Merge pull request #10425 from dshoreman/ds/nordigen-tx-days

Nordigen Transaction history UI

turbo124 committed:
Updates for composer dependencies

ITSYV committed:
Purify HTML: fix tagline

turbo124 committed:
Merge pull request #10483 from ITSYV/v5-develop

Purify HTML: fix tagline

turbo124 committed:
v5.11.12

turbo124 committed:
Merge pull request #10485 from turbo124/v5-develop

v5.11.12

turbo124 committed:
Fixes for PDF's

turbo124 committed:
Merge pull request #10486 from turbo124/v5-develop

Fixes for PDF's

turbo124 committed:
Fixes for pdf

turbo124 committed:
Reversion for encoded html fixes

turbo124 committed:
v5.11.13

turbo124 committed:
Merge pull request #10487 from turbo124/v5-develop

v5.11.13

turbo124 committed:
v5.11.14:

turbo124 committed:
Merge pull request #10488 from turbo124/v5-develop

v5.11.14:

turbo124 committed:
fixes for emailing all contacts

turbo124 committed:
Merge pull request #10489 from turbo124/v5-develop

Fixes for emailing all contacts

turbo124 committed:
minor fix for settings prop check

turbo124 committed:
Merge pull request #10491 from turbo124/v5-develop

Minor fix for settings prop check

turbo124 committed:
Updates for composer dependencies

benbrummer committed:
Create an additional archive with tar.gz and toplevel directory

Signed-off-by: benbrummer <info@benjamin-brummer.de>

turbo124 committed:
Merge pull request #10475 from benbrummer/v5-develop

react_release.yml improvements

turbo124 committed:
additional props for invoice items

turbo124 committed:
Fixes for unicode characters in html

turbo124 committed:
Merge pull request #10492 from turbo124/v5-develop

Fixes for unicode characters in html

turbo124 committed:
Fixes for react build

turbo124 committed:
v5.11.16

turbo124 committed:
Merge pull request #10493 from turbo124/v5-develop

v5.11.16

turbo124 committed:
Fixes for react build

turbo124 committed:
v5.11.17

turbo124 committed:
Merge pull request #10494 from turbo124/v5-develop

v5.11.17

benbrummer committed:
Exclude previously created invoiceninja.tar for invoiceninjar.tar.gz

Signed-off-by: benbrummer <info@benjamin-brummer.de>

benbrummer committed:
Merge branch 'invoiceninja:v5-develop' into v5-develop

benbrummer committed:
mv is wrong here

Signed-off-by: benbrummer <info@benjamin-brummer.de>

turbo124 committed:
Merge pull request #10495 from benbrummer/v5-develop

do not mv invoiceninjar.tar.gz into invoiceninja folder

turbo124 committed:
Fixes for invoice terms display incorrectly

turbo124 committed:
Fixes for typo in test

turbo124 committed:
Fixes for company imports

turbo124 committed:
Fixes for bcc on payment failure mails

turbo124 committed:
Clean up for HTML purifier

turbo124 committed:
Fixes for static analysis

turbo124 committed:
Adjustments for importing customers with auth.net

turbo124 committed:
minor fixes

turbo124 committed:
Fixes for peppol routing/tax scheme

turbo124 committed:
v5.11.18

turbo124 committed:
Merge pull request #10504 from turbo124/v5-develop

v5.11.18

turbo124 committed:
Hide shipping address if none is configured

turbo124 committed:
Cleanup

turbo124 committed:
Minor fixes for storecove testS

turbo124 committed:
Revert latest nordigen commit

turbo124 committed:
Fixes for QR codes in PDFs

turbo124 committed:
Fixes for pdf conversion issues and hidden SVG elements

turbo124 committed:
v5.11.19

turbo124 committed:
Merge pull request #10506 from turbo124/v5-develop

v5.11.19

turbo124 committed:
Fixes for division by zero

turbo124 committed:
v5.11.20

turbo124 committed:
Merge pull request #10510 from turbo124/v5-develop

v5.11.20

turbo124 committed:
Minor Fixes

turbo124 committed:
v5.11.21

turbo124 committed:
Merge pull request #10511 from turbo124/v5-develop

v5.11.21

paulwer commented:
@dshoreman we may should considder omitting reusing existing agreements at all to reduce complexity.

dshoreman commented:
Paul please, just give me a chance to check. This is already getting frustrating.

benbrummer committed:
Integrate octane in preload.php

Signed-off-by: benbrummer <info@benjamin-brummer.de>

benbrummer committed:
comment echo

reduce noise

Signed-off-by: benbrummer <info@benjamin-brummer.de>

turbo124 committed:
Fixes for xinvoice attachments

turbo124 committed:
v5.11.22

turbo124 committed:
Merge pull request #10518 from turbo124/v5-develop

v5.11.22

benbrummer committed:
added classes to ignore list again, execpt \Illuminate\Http\Testing\File::class,

Signed-off-by: benbrummer <info@benjamin-brummer.de>

turbo124 committed:
Adjustments for design and dom elements / attributes

turbo124 committed:
Fixes for delivery notes

turbo124 committed:
v5.11.23

turbo124 committed:
Merge pull request #10521 from turbo124/v5-develop

Fixes for show/hide elements

turbo124 committed:
Update invoice event when matching on invoice transactions

turbo124 committed:
Merge pull request #10517 from benbrummer/v5-develop

Integrate octane in preload.php

turbo124 committed:
Merge pull request #10522 from turbo124/v5-develop

Update invoice event when matching on invoice transactions

dshoreman committed:
Fix rebase/refactor typos

Fixes the bank buttons which weren't getting their handlers registered
thanks to the `null` making it go down the wrong path, and makes sure to
use the institution id from the object passed in, not the institution_id
variable from the global scope which is only defined for renewals.

Resolves #10501
